### PR TITLE
ViewModelInterface

### PR DESCRIPTION
**Pull request by oguzkocer**
**TL;DR:** This PR is about how we can solve common problems for ViewModels in a consistent way. It talks about a specific problem, activity re-creation, and proposes a solution for it. It's meant to be the first step of documenting the `ViewModel` pattern to make it easier for any developer to utilize in a consistent manner.

--------------------------------

We have only recently started using `ViewModel`s in the app with the introduction of `PluginBrowserViewModel`. That gives us a great opportunity to document everything properly from the start and introduce a good structure for all of us to follow. Although there will be a document in the repository to explain and showcase the patterns we use, I strongly believe that the code should be self-documenting as much as possible. By that I mean, instead of saying we should do `X` in a document and bring it up during reviews, the structure itself should enforce doing that and ideally explain why and how of it as a javadoc. Once we figure out a solution for a common problem and settle on it, we shouldn't have to re-solve it unless something has changed that makes the solution invalid (not-ideal).

This PR is only the first step for this. I'll try to contain each decision/improvement in a PR like this one, so we can make incremental decisions and document them along the way.

**The Problem**

> ViewModels only survive configuration change-related destruction; they do not survive the process being stopped. This makes ViewModels a replacement for using a fragment with setRetainInstance(true) (in fact ViewModels use a fragment with setRetainInstance set to true behind the scenes). [#](https://medium.com/google-developers/viewmodels-persistence-onsaveinstancestate-restoring-ui-state-and-loaders-fc7cc4a6c090)

What this means is that, we don't need to handle configuration changes in our ViewModel, the data you have cached in your ViewModel will survive these which is very important. However, if the process is stopped, we still need a way to re-create the Activity/Fragment. In practice, this means we still need to keep the `id`s of the objects we are interested in, so that if the ViewModel is re-created we can ask FluxC (or another data source) for the data.

_This is all good, but how do we do that and why is it not straightforward?_

The activity does not know about whether it was re-created due to configuration change or because the process has stopped, so it can't tell the `ViewModel` to do an `initial setup*`. The `ViewModel` can not do the initial setup in its constructor either, because in most, if not all, cases it'll need some kind of information from the Activity and since the Activity does not handle the construction of the ViewModel.

Although there are a lot of different ways to go about this, the only ones I am comfortable with that I've thought of so far are:
1. Use an interface for the common methods that each ViewModel will need to implement
2. Use an abstract class for the same reason

If there are alternatives, I'll always prefer NOT TO add a base class for pretty much anything. I think the **implementation inheritance** is an evil concept and should be avoided like a plague. However, my personal opinion of them is not important and all it matters is that we are **consistent** about it.

I am not going to document the implementation details of the `ViewModelInterface` here, because I think the documentation in the code should do that. If it's not good enough and if anything is not clear, we should just improve on that instead of explaining here.

**Initial Setup**
I don't know if this was clear from the context, but these are the things we want to do only once when the `ViewModel` is created but can't be done in the constructor. Let's say the `ViewModel` needs the `SiteModel` to be able to fetch/load plugins for that site. Since we don't know what that site is in the constructor, we can't do it there, so we need to do that in a different place. (`onStart` naming comes from a Google example or doc, I can't remember it exactly, but Google was using the `start` method)

**Expectation**

I really think every Android developer at a8c should at least check this PR out. I expect these to be the foundation of the app going forward. Here are my suggestion of steps everyone should consider (without skipping any):

1. Make sure to understand why `ViewModel`s might be re-created.
2. Check out the current approach of how the initial setup happens currently (`onStarted`, `writeToBundle`, `readFromBundle`). If you have a different idea of how we might do this, please make sure to share that.
3. **If you agree with the current approach for the initial setup** check out the current `Interface` approach and consider whether the `abstract class` approach would have been better and why. I only suggested 2 ways, but please share any other approach you think might work.
4. **If you agree with everything so far**, it'd be great if you can read the documentation in `ViewModelInterface` and give feedback on how it can be improved. (**Bonus**)

_P.S:_ If you give thumbs up to this PR, I'll consider that you completed the first 3 steps and you are in agreement with it. Also, please let me know if you have any feedback on how these proposals can be done in a better way.

### CODE REVIEW

**Comment by malinajirka** [comment]
This is great @oguzkocer, thanks! 

I'd vote for the `abstract class` approach. We wouldn't need to handle `isStarted` state in each VM -> we'd save some code and it would be less error prone. 
It's a bad idea to use implementation inheritance in a library when you want to add a specific feature to a class - we all know the evil of `SherlockActivity`, `SherlockMapActivity`, `DaggerActivity`, `AppCompatActivity`, `SlidingMenuActivity` etc. However, I don't see any issues with implementation inheritance in this particular case.

**Comment by nbradbury** [comment]
First off, thanks for taking the lead in researching `ViewModels`. I know many of us have been wanting to dive into them.

I do get why implementation inheritance can go wrong, but in most cases I prefer to avoid blanket statements about an approach being "evil." Sometimes a choice that has side effects in one situation works well for another. For example, using `setRetainInstance(true)` in fragments often gets a bad rap, yet it's the backbone of `ViewModels`.

I tend to agree with @malinajirka that an `abstract class` approach would work fine here.

One red flag I'd like to raise: `ViewModels` are Google's proposed solution to the issue of retaining state after configuration changes, yet already we're discovering this solution has [pitfalls](https://medium.com/@BladeCoder/architecture-components-pitfalls-part-1-9300dd969808) of it's own. It makes me wonder whether we need to let this solution mature for a little bit so we don't end up writing code to tackle pitfalls that Google will address themselves.

**Comment by oguzkocer** [comment]
> I do get why implementation inheritance can go wrong, but in many cases I prefer to avoid blanket statements about an approach being "evil"

I clearly marked that as a personal opinion, one that does not matter in fact, so calling it a "blanket statement" feels a bit harsh to me. I specifically avoided talking too much about this, and only mentioned my take on it, because I didn't want the discussion to be around inheritance. Using an `abstract class` for this is definitely a valid option, my only concern is that we start adding more things to this abstract class, when we shouldn't, because it's the easier option. I believe in patterns that help prevent developer errors as much as possible and I consider base classes, even if `abstract`, a dangerous pattern because not only it doesn't prevent them, it actually encourages it. Again, my take on this is NOT IMPORTANT (not to me anyway) especially because the `Interface` pattern here is not ideal either. I care that we are consistent in our solution and I'll do my best to write a good documentation around it to help avoid those kinds of mistakes.

> For example, using `setRetainInstance(true)` in fragments often gets a bad rap, yet it's the backbone of ViewModels.

My understanding is that using `setRetainInstance(true)` fragment was the most feasible option for the Google to go with since this is being added several years after Android is designed to help fix an issue they themselves has created. To me, that does not mean that it was a good option, it just means that it was the best option.

I'd say this whole PR is about the poor design behind configuration changes and Google's previous lackluster solution of `setRetainInstance` actually. If they did a better job with the design initially, we wouldn't have had to deal with this issue in a custom way like this (which they could have helped with as well... )

> One red flag I'd like to raise: `ViewModels` are Google's proposed solution to the problems of retaining state after configuration changes, yet already we're discovering this solution has pitfalls of it's own.

I read that article before I started implementing the `ViewModel`, but I haven't had a chance to fully explore it, so I am not sure how valid of a concern it is. It's important to note that the issue is not actually about the `ViewModel` but it's about the `LiveData`. `LiveData` compared to other type of observables is life-cycle aware, which is a bonus and not a requirement, and all this article says is that this bonus can go away in some circumstances which is something we can handle fairly easily. It's not like we don't have these same issues (and more) right now. Most FluxC events goes bye-bye during activity re-creation, config changes etc. At least we are able to handle those easily in `ViewModel` right now.

I will explore this a bit more and make sure it's not a problem for us and the solution is properly documented as well, assuming that we need one in the first place.

> It makes me wonder whether we need to let this solution mature for a little bit so we don't end up writing code to tackle pitfalls that Google will address themselves.

I don't think that's necessary. As long as our patterns are properly documented, if Google makes improvements/changes, we should be able to react to them easily.

**Comment by nbradbury** [comment]
> I clearly marked that as a personal opinion, one that does not matter in fact, so calling it a "blanket statement" feels a bit harsh to me

I did not intend that as a personal attack and apologize if it came off that way. Let's drop the matter and move on.

> To me, that does not mean that it was a good option, it just means that it was the best option.

Yes, exactly right!

> I read that article before I started implementing the ViewModel, but I haven't had a chance to fully explore it, so I am not sure how valid of a concern it is

I should clarify that by linking to that article I didn't mean to suggest that the issues it raised were valid - I haven't researched it enough to know - but instead to give one data point to illustrate how developers are already finding issues and coding workarounds for `ViewModel`. As you know, there are plenty of other similar articles.

> I don't think that's necessary. As long as our patterns are properly documented, if Google makes improvements/changes, we should be able to react to them easily.

This sounds good to me. Truth be told, I'm stuck in a loop of wanting to try out new Android features as soon as they come out, but being wary of them because we're been burned before (hello support library!).

FTR, I think the `ViewModel` approach is better than the current mess of saving/restoring state, especially since the "current mess" doesn't scale for storing more than small bits of data.

**Comment by hypest** [comment]
Thanks for initiating a discussion on this @oguzkocer !

While I'm still trying to wrap my head around the issues and the proposals, let me ask:

> The activity does not know about whether it was re-created due to configuration change or because the process has stopped...

I think the Activity should not need to know or assume the state of the VM at the time the Activity is (re)created. My impression is that as the AAC classes (LiveData in particular) are pushing towards a more reactive model of composing the classes, the Activity should just need to subscribe to a VM's state. Based on that state (the observer receives it upon subscribing anyway) the Activity will respond with priming the VM if the VM is in, say, an `INIT` state. In essence, the Activity will "detect" the state of the VM and act accordingly without having to know if a configuration change or process stop happened. Does that make any sense? Am I perhaps missing something?

**Comment by oguzkocer** [comment]
@hypest The activity doesn't need to know about the state of the VM or how it's re-created (due to config changes or process stop). The VM does need to know whether an initial setup is necessary or not.

> The activity does not know about whether it was re-created due to configuration change or because the process has stopped...

All that means is that we can not pass that information to the VM.

Hope that clears it up!

**Comment by malinajirka** [comment]
I have been thinking about it and I have one more question.

Do we really need the `writeToBundle(..)` and `readFromBundle(..)` methods?

When I want to view let's say a post detail, I start PostDetailActivity and I pass the post `id` via an Intent. When the activity is killed (actually the whole process is killed) and the user returns back to the application, the Android Framework starts(recreates) the activity with the same Intent as it was started before, right? So the post id is still there. From the viewModel's point of view this is exactly the same scenario as if it was started for the first time.

I'm wondering whether this is not the reason why has Google decided to design viewModels without `writeToBundle(..)` and `readFromBundle(..)` methods. I've always tried to keep the ViewModels Android Framework independent as much as possible. WDYT?

**Comment by 0nko** [comment]
Good discussion, @oguzkocer! I recently read a HN post about [inheritance](https://www.sicpers.info/2018/03/why-inheritance-never-made-any-sense), where the author talks about 3 reasons for using inheritance (subtyping, abstraction and code sharing). Although he cautions against using inheritance as a solution for everything, it may make sense when used for a single purpose. I'd say code sharing could be considered such purpose here to warrant the use of a base class.

Another thing that ocurred to me when reading the `ViewModelInterface` was that there are a couple of assumptions made on the Activity side. For example:
- "The activity should be calling this method in its `onCreate`"
- "This method should be called before `onStart` in the activity!"
- "The activity should be calling this method from its onCreate(Bundle savedInstanceState) method only if the `savedInstanceState` is not `null`."

Perhaps some ViewModelActivity base class might be useful, which would make sure all these requirements are satisfied. There is no way to enforce it otherwise and it's easy to forget.

Finally, I also wondered about this:
> Do we really need the writeToBundle(..) and readFromBundle(..) methods?

When the activity is destroyed along with a ViewModel, can't we simply restore it by recreating the ViewModel from scratch, just like when it's initially started?

**Comment by oguzkocer** [comment]
@malinajirka @0nko Unfortunately we can't use the `Intent` to re-create the `ViewModel` in some cases. For example, there is a search view in the Plugin Browser page and the search query is something we need to keep in the bundle, because it's not something passed from somewhere else. So, at least for some things, we definitely need it.

@0nko I need to think about your suggestion for base activity a bit more, but I don't think it's necessary and I don't think it'll be pretty because we'll need information from the child class about which `ViewModel` we would like to create as well as give it a way to pass the necessary information (such as the site) to the ViewModel. Even without it, I am not too worried that it'll be forgotten because it'll be part of the `ViewModel` creation in the Activity and we'll have code examples around it.

Couple of notes about the discussion so far:
1. This is super helpful, because I can see what other people's concerns are and maybe come up with something that makes the most (hopefully all) of us happy
2. I have not considered solutions that could utilize Kotlin. I'd like to convert the existing ViewModel we have to Kotlin and document everything in it. Maybe there is some better pattern we can utilize with Kotlin. I'll explore that soon, let me know if there is anything you'd like me to try.

Thanks everyone for your comments so far, please keep them coming!

**Comment by malinajirka** [comment]
>@malinajirka @0nko Unfortunately we can't use the Intent to re-create the ViewModel in some cases. For example, there is a search view in the Plugin Browser page and the search query is something we need to keep in the bundle, because it's not something passed from somewhere else. So, at least for some things, we definitely need it.

I believe the Android Framework takes care of restoring view state (if the view has an `id`) in such cases. Or am I missing something?

**Comment by oguzkocer** [comment]
> I believe the Android Framework takes care of restoring view state (if the view has an id) in such cases.

I guess the search query was a poor example. SearchView and EditText does retain their state if I am not mistaken during config changes, but I am not sure whether they still do so after process is completely stopped. Whether it does is not very important since this is not the main point of this discussion.

There could be many other cases, such as a dialog being shown, selected items in a list, whether more items can be loaded for a list (which unfortunately we don't keep in FluxC right now), state of a multi-step request (such as chaining multiple FluxC actions), so on. Some activities might not require it and we could definitely use the `Intent` for those, but for some we'll definitely need to save things in the Bundle.

I've actually linked [an article by Google](https://medium.com/google-developers/viewmodels-persistence-onsaveinstancestate-restoring-ui-state-and-loaders-fc7cc4a6c090) in my original PR description which talks about the need for `onSaveInstanceState` which I suggest everyone to read if you haven't already.

Hope that clears up the confusion, let me know if you have further questions about this!

**Comment by oguzkocer** [comment]
_Quick note on the progress on this:_ I've decided to go with the Kotlin conversion before continuing the work on the documentation. I have a couple ideas I want to try out that might help us get rid of the `onStart` method. We'll still need to do something similar, but in a less strict & error-prone way.

I'll write another update here when they are done. In the meantime, please feel free to check out the Kotlin conversion PR #7551 and other improvements and give feedback on them. Let me know if you want me to ping you on the upcoming PRs.

Thanks again to everyone for your input so far!

**Comment by oguzkocer** [comment]
It turns out that we don't need as much structure for all ViewModels as I initially expected. We'll probably still have some common things to handle, but nothing important enough to force a specific way to handle it (and definitely not enough to introduce base classes). Especially with Kotlin, things should be a lot easier to handle.

I am going to close this PR and write a p2 post about some suggestions and gotchas about ViewModel after I am done with the Zendesk integration. Thanks again for the input in this PR.
